File: runtime/android/app/src/main/java/com/appunvs/runtime/ui/BoxSwitcher.kt
// BoxSwitcher — Chat header chip + ModalBottomSheet listing boxes.
// Tapping the chip opens a sheet with: the box list (active marked
// with a brand check), a "+ New box" row at the bottom, and a
// disabled "Scan QR" placeholder for the connector flow.
package com.appunvs.runtime.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AddCircle
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material3.Divider
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch

import com.appunvs.runtime.state.Box
import com.appunvs.runtime.state.BoxState
import com.appunvs.runtime.state.MockStore
import com.appunvs.runtime.theme.LocalAppColors
import com.appunvs.runtime.theme.Radius
import com.appunvs.runtime.theme.Spacing

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BoxSwitcher(
    store: MockStore,
    onNewBox: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = LocalAppColors.current
    var sheetOpen by remember { mutableStateOf(false) }
    val sheetState = rememberModalBottomSheetState()
    val scope = rememberCoroutineScope()

    Row(
        modifier = modifier
            .clip(RoundedCornerShape(Radius.m.dp))
            .clickable { sheetOpen = true }
            .padding(horizontal = Spacing.s.dp, vertical = Spacing.xs.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = store.activeBox?.title ?: "选择 Box",
            style = MaterialTheme.typography.bodyLarge.copy(
                color = colors.textPrimary,
                fontWeight = FontWeight.SemiBold,
            ),
            maxLines = 1,
            modifier = Modifier.widthIn(max = 220.dp),
        )
        Icon(
            imageVector = Icons.Filled.ArrowDropDown,
            contentDescription = null,
            tint = colors.textSecondary,
        )
    }

    if (sheetOpen) {
        ModalBottomSheet(
            onDismissRequest = { sheetOpen = false },
            sheetState = sheetState,
            containerColor = colors.bgCard,
        ) {
            Column(modifier = Modifier.padding(bottom = Spacing.xl.dp)) {
                Text(
                    text = "我的 Box",
                    style = MaterialTheme.typography.titleMedium.copy(
                        color = colors.textPrimary,
                        fontWeight = FontWeight.Bold,
                    ),
                    modifier = Modifier.padding(horizontal = Spacing.l.dp, vertical = Spacing.s.dp),
                )

                store.boxes.forEach { box ->
                    BoxRow(
                        box = box,
                        isActive = box.id == store.activeBox?.id,
                        onTap = {
                            store.setActive(box)
                            scope.launch {
                                sheetState.hide()
                                sheetOpen = false
                            }
                        },
                    )
                    Divider(color = colors.borderDefault)
                }

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable {
                            scope.launch {
                                sheetState.hide()
                                sheetOpen = false
                                onNewBox()
                            }
                        }
                        .padding(horizontal = Spacing.l.dp, vertical = Spacing.m.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        imageVector = Icons.Filled.AddCircle,
                        contentDescription = null,
                        tint = colors.brandDark,
                    )
                    Spacer(Modifier.width(Spacing.m.dp))
                    Text(
                        text = "新建 Box",
                        style = MaterialTheme.typography.bodyLarge.copy(
                            color = colors.textPrimary,
                            fontWeight = FontWeight.SemiBold,
                        ),
                    )
                }

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = Spacing.l.dp, vertical = Spacing.m.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        imageVector = Icons.Filled.QrCodeScanner,
                        contentDescription = null,
                        tint = colors.textSecondary,
                    )
                    Spacer(Modifier.width(Spacing.m.dp))
                    Text(
                        text = "扫码看别人的 app",
                        style = MaterialTheme.typography.bodyLarge.copy(
                            color = colors.textSecondary,
                            fontWeight = FontWeight.SemiBold,
                        ),
                    )
                    Spacer(Modifier.weight(1f))
                    AppBadge("即将上线", tone = BadgeTone.NEUTRAL)
                }
            }
        }
    }
}

@Composable
private fun BoxRow(
    box: Box,
    isActive: Boolean,
    onTap: () -> Unit,
) {
    val colors = LocalAppColors.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onTap() }
            .background(if (isActive) colors.brandPale else colors.bgCard)
            .padding(horizontal = Spacing.l.dp, vertical = Spacing.m.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = box.title,
                style = MaterialTheme.typography.bodyLarge.copy(
                    color = colors.textPrimary,
                    fontWeight = FontWeight.SemiBold,
                ),
                maxLines = 1,
            )
            Text(
                text = "v${box.currentVersion.ifEmpty { "—" }}",
                style = MaterialTheme.typography.bodySmall.copy(color = colors.textSecondary),
            )
        }
        AppBadge(box.state.label, tone = box.state.tone)
        if (isActive) {
            Spacer(Modifier.width(Spacing.s.dp))
            Icon(
                imageVector = Icons.Filled.Check,
                contentDescription = "active",
                tint = colors.brandDark,
            )
        }
    }
}

private val BoxState.label: String
    get() = when (this) {
        BoxState.DRAFT     -> "draft"
        BoxState.PUBLISHED -> "published"
        BoxState.ARCHIVED  -> "archived"
    }

private val BoxState.tone: BadgeTone
    get() = when (this) {
        BoxState.DRAFT     -> BadgeTone.WARNING
        BoxState.PUBLISHED -> BadgeTone.SUCCESS
        BoxState.ARCHIVED  -> BadgeTone.NEUTRAL
    }

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NewBoxSheet(
    onDismiss: () -> Unit,
    onCreate: (String) -> Unit,
) {
    val colors = LocalAppColors.current
    val sheetState = rememberModalBottomSheetState()
    val scope = rememberCoroutineScope()
    var title by remember { mutableStateOf("") }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = colors.bgCard,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(Spacing.l.dp),
            verticalArrangement = Arrangement.spacedBy(Spacing.m.dp),
        ) {
            Text(
                text = "新建 Box",
                style = MaterialTheme.typography.titleMedium.copy(
                    color = colors.textPrimary,
                    fontWeight = FontWeight.Bold,
                ),
            )
            androidx.compose.material3.OutlinedTextField(
                value = title,
                onValueChange = { title = it },
                placeholder = { Text("比如 todo-app") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
            )
            Text(
                text = "Box 是一个独立项目, 对话历史和源代码都和它绑定。",
                style = MaterialTheme.typography.bodySmall.copy(color = colors.textSecondary),
            )
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(Spacing.s.dp),
            ) {
                androidx.compose.material3.OutlinedButton(
                    onClick = {
                        scope.launch {
                            sheetState.hide()
                            onDismiss()
                        }
                    },
                    modifier = Modifier.weight(1f),
                ) { Text("取消") }
                androidx.compose.material3.Button(
                    onClick = {
                        val t = title.trim()
                        if (t.isNotEmpty()) {
                            onCreate(t)
                            scope.launch {
                                sheetState.hide()
                                onDismiss()
                            }
                        }
                    },
                    enabled = title.trim().isNotEmpty(),
                    modifier = Modifier.weight(1f),
                ) { Text("创建") }
            }
        }
    }
}

File: appunvs/ios/Runtime/UI/BoxSwitcher.swift
// BoxSwitcher — Chat header chip + bottom sheet listing boxes.
// Tapping the chip opens a sheet with the box list (current marked
// with a brand check), and footer actions for "+ New box" and
// "Scan QR" (the latter still disabled until the pairing flow lands).
import SwiftUI

struct BoxSwitcher: View {
    @EnvironmentObject private var store: BoxStore
    @State private var sheetOpen: Bool = false
    @State private var newBoxOpen: Bool = false

    var body: some View {
        Button {
            sheetOpen = true
        } label: {
            HStack(spacing: Spacing.xs) {
                Text(store.activeBox?.title ?? "选择 Box")
                    .font(.body.weight(.semibold))
                    .foregroundStyle(Theme.textPrimary.color)
                    .lineLimit(1)
                Image(systemName: "chevron.down")
                    .font(.caption)
                    .foregroundStyle(Theme.textSecondary.color)
            }
            .padding(.horizontal, Spacing.s)
            .padding(.vertical, Spacing.xs)
            .background(
                RoundedRectangle(cornerRadius: Radius.m)
                    .fill(Color.clear)
            )
        }
        .buttonStyle(.plain)
        .sheet(isPresented: $sheetOpen) {
            BoxSwitcherSheet(
                onCreate: {
                    sheetOpen = false
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                        newBoxOpen = true
                    }
                },
                onSelect: { box in
                    store.setActive(box)
                    sheetOpen = false
                }
            )
            .presentationDetents([.medium, .large])
            .presentationDragIndicator(.visible)
        }
        .sheet(isPresented: $newBoxOpen) {
            NewBoxSheet { title in
                Task { await store.create(title: title) }
                newBoxOpen = false
            }
        }
    }
}

private struct BoxSwitcherSheet: View {
    @EnvironmentObject private var store: BoxStore
    let onCreate: () -> Void
    let onSelect: (BoxWire) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Text("我的 Box")
                    .font(.title3.weight(.bold))
                    .foregroundStyle(Theme.textPrimary.color)
                Spacer()
                Button {
                    Task { await store.refresh() }
                } label: {
                    Image(systemName: "arrow.clockwise")
                        .font(.body)
                        .foregroundStyle(Theme.textSecondary.color)
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, Spacing.l)
            .padding(.top, Spacing.m)
            .padding(.bottom, Spacing.s)

            ScrollView {
                if store.boxes.isEmpty {
                    Text(store.loading ? "加载中…" : "还没有 Box, 点下方新建一个")
                        .font(.subheadline)
                        .foregroundStyle(Theme.textSecondary.color)
                        .frame(maxWidth: .infinity, alignment: .center)
                        .padding(Spacing.xxl)
                } else {
                    VStack(spacing: 0) {
                        ForEach(store.boxes) { box in
                            BoxRow(
                                box: box,
                                isActive: box.boxID == store.activeBox?.boxID,
                                onTap: { onSelect(box) }
                            )
                            Divider().background(Theme.borderDefault.color)
                        }
                    }
                }
            }

            Button(action: onCreate) {
                HStack(spacing: Spacing.m) {
                    Image(systemName: "plus.circle.fill")
                        .font(.title2)
                        .foregroundStyle(Theme.brandDark.color)
                    Text("新建 Box")
                        .font(.body.weight(.semibold))
                        .foregroundStyle(Theme.textPrimary.color)
                    Spacer()
                }
                .padding(.horizontal, Spacing.l)
                .padding(.vertical, Spacing.m)
            }
            .buttonStyle(.plain)

            HStack(spacing: Spacing.m) {
                Image(systemName: "qrcode.viewfinder")
                    .font(.title2)
                    .foregroundStyle(Theme.textSecondary.color)
                Text("扫码看别人的 app")
                    .font(.body.weight(.semibold))
                    .foregroundStyle(Theme.textSecondary.color)
                Spacer()
                Badge("即将上线", tone: .neutral)
            }
            .padding(.horizontal, Spacing.l)
            .padding(.vertical, Spacing.m)
            .opacity(0.5)
        }
        .background(Theme.bgCard.color)
    }
}

private struct BoxRow: View {
    let box: BoxWire
    let isActive: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            HStack(spacing: Spacing.m) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(box.title)
                        .font(.body.weight(.semibold))
                        .foregroundStyle(Theme.textPrimary.color)
                        .lineLimit(1)
                    Text("v\(box.currentVersion.isEmpty ? "—" : box.currentVersion)")
                        .font(.caption)
                        .foregroundStyle(Theme.textSecondary.color)
                }
                Spacer()
                Badge(box.state.rawValue, tone: tone(for: box.state))
                if isActive {
                    Image(systemName: "checkmark")
                        .font(.caption.weight(.bold))
                        .foregroundStyle(Theme.brandDark.color)
                }
            }
            .padding(.horizontal, Spacing.l)
            .padding(.vertical, Spacing.m)
            .background(isActive ? Theme.brandPale.color : Color.clear)
        }
        .buttonStyle(.plain)
    }

    private func tone(for state: BoxStateWire) -> BadgeTone {
        switch state {
        case .draft:       return .warning
        case .published:   return .success
        case .archived:    return .neutral
        case .unspecified: return .neutral
        }
    }
}

private struct NewBoxSheet: View {
    let onCreate: (String) -> Void
    @Environment(\.dismiss) private var dismiss
    @State private var title: String = ""

    var body: some View {
        NavigationStack {
            Form {
                Section("名称") {
                    TextField("比如 todo-app", text: $title)
                        .autocorrectionDisabled()
                        .textInputAutocapitalization(.never)
                }
                Section {
                    Text("Box 是一个独立项目, 对话历史和源代码都和它绑定。")
                        .font(.footnote)
                        .foregroundStyle(Theme.textSecondary.color)
                }
            }
            .scrollContentBackground(.hidden)
            .background(Theme.bgPage.color)
            .navigationTitle("新建 Box")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("取消") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("创建") {
                        let t = title.trimmingCharacters(in: .whitespaces)
                        guard !t.isEmpty else { return }
                        onCreate(t)
                    }
                    .disabled(title.trimmingCharacters(in: .whitespaces).isEmpty)
                }
            }
        }
        .presentationDetents([.medium])
    }
}
